site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+---
+title: "OpenTracing for Go Projects"
+slug: open-tracing
+series: "AST"
+date: 2020-06-10T00:00:00-04:00
+author: Eli Yukelzon
+github: reflog
+community: eli.yukelzon
+toc: true
+---
+
+## What is distributed tracing?
+
+Large scale, cloud applications are usually built using interconnected services are rather hard to troubleshoot. When a service is scaled, simple logging doesn't cut it anymore and a more in-depth view into system's flow is required.
+That's where [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/) comes into play: it allows developers and SREs to get a detailed view of a request as it travels through the system of services. With distributed tracing you can:
+
+1. Trace the execution path of a single request as it goes through a complicated path inside the distributed system
+2. Pinpoint bottlenecks and measure latency of specific parts of the execution path
+3. Record and analyze system behavior
+
+[OpenTracing](https://opentracing.io) is an open standard describing how the way distributed tracing works.
+
+There are a few key terms used in tracing:
+
+1. **Trace** - a recording of the execution path of a request
+2. **Span** - a named, timed operation representing a contiguous segment inside the trace
+3. **Root Span** - the first span in a trace, a common ancestor to all spans in a trace
+4. **Context** - information identifying the request, required to connect spans in a distributed trace
+
+A trace recording usually looks something like this:
+
+![trace image](/blog/2020-06-10-opentracing/trace.png) 
+
+To add distributed tracing capabilities to [mattermost-server](https://github.com/mattermost/mattermost-server) we've picked [OpenTracing Go](https://github.com/opentracing/opentracing-go). We've already dove a little bit into Open Tracing implementation in the [first post of these series]({{< relref "./2019-10-25-instrumenting-go-code-via-ast.md" >}}).
+
+In this article we'll discuss all the nitty-gritty details of implementing a tracing system in your Go application without littering your code with repetitive, boilerplate tracing code.
+
+## The Goal
+
+So what are we actually working on? We want to make sure that every API request that is being handled by our server will get recorded into a trace, together with context information and ability to dive deep into the execution and allow easy problem analysis.
+
+The resulting system trace will look like this (using [Jaeger](https://www.jaegertracing.io/) web-ui visualization):
+![jaeger view](/blog/2020-06-10-opentracing/result.png) 
+
+## Naive approach
+
+To add tracing to any API call, we can do the following in our `ServeHTTP` function:
+
+```go
```

Comment:
Might be also good to add the import statements so that it becomes clear what packages are imported.